[Bug]: multi_process_shared.py doesn't seem to correctly release models

### What happened?

When setting always_proxy=True in multi_process_shared.py, releasing an acquired model causes the following exception:

```
apache_beam/ml/inference/base_test.py:1023: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
apache_beam/ml/inference/base.py:788: in load
    shared_handle.release(model_to_remove)
apache_beam/utils/multi_process_shared.py:255: in release
    self._manager.release_singleton(self._tag, obj)
../../../.pyenv/versions/3.8.9/lib/python3.8/multiprocessing/managers.py:740: in temp
    token, exp = self._create(typeid, *args, **kwds)
../../../.pyenv/versions/3.8.9/lib/python3.8/multiprocessing/managers.py:625: in _create
    id, exposed = dispatch(conn, None, 'create', (typeid,)+args, kwds)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

c = <multiprocessing.connection.Connection object at 0x1244356a0>, id = None, methodname = 'create'
args = ('release_singleton', '138c832ae04d4c1a9505cdfb1382d630', <apache_beam.utils.multi_process_shared._AutoProxyWrapper object at 0x123fb1550>)
kwds = {}

    def dispatch(c, id, methodname, args=(), kwds={}):
        '''
        Send a message to manager using connection `c` and return response
        '''
        c.send((id, methodname, args, kwds))
        kind, result = c.recv()
        if kind == '#RETURN':
            return result
>       raise convert_to_error(kind, result)
E       multiprocessing.managers.RemoteError: 
E       ---------------------------------------------------------------------------
E       Traceback (most recent call last):
E         File "/Users/dannymccormick/.pyenv/versions/3.8.9/lib/python3.8/multiprocessing/managers.py", line 202, in handle_request
E           request = c.recv()
E         File "/Users/dannymccormick/.pyenv/versions/3.8.9/lib/python3.8/multiprocessing/connection.py", line 251, in recv
E           return _ForkingPickler.loads(buf.getbuffer())
E         File "/Users/dannymccormick/beam/sdks/python/apache_beam/utils/multi_process_shared.py", line 152, in __getattr__
E           return getattr(self._proxyObject, name)
E         File "/Users/dannymccormick/beam/sdks/python/apache_beam/utils/multi_process_shared.py", line 152, in __getattr__
E           return getattr(self._proxyObject, name)
E         File "/Users/dannymccormick/beam/sdks/python/apache_beam/utils/multi_process_shared.py", line 152, in __getattr__
E           return getattr(self._proxyObject, name)
E         [Previous line repeated 991 more times]
E       RecursionError: maximum recursion depth exceeded
E       ---------------------------------------------------------------------------

../../../.pyenv/versions/3.8.9/lib/python3.8/multiprocessing/managers.py:91: RemoteError
```

### Issue Priority

Priority: 2 (default / most bugs should be filed as P2)

### Issue Components

- [X] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

--- a/sdks/python/apache_beam/utils/multi_process_shared.py
+++ b/sdks/python/apache_beam/utils/multi_process_shared.py
@@ -24,8 +24,6 @@ on it via rpc.
 
 import logging
 import multiprocessing.managers
-from inspect import signature
-from functools import wraps
 import os
 import tempfile
 import threading

--- a/sdks/python/apache_beam/utils/multi_process_shared.py
+++ b/sdks/python/apache_beam/utils/multi_process_shared.py
@@ -46,7 +46,6 @@ autoproxy = multiprocessing.managers.AutoProxy
 
 def patched_autoproxy(token, serializer, manager=None, authkey=None,
           exposed=None, incref=True, manager_owned=True):
-    # Calling original AutoProxy without the unwanted key argument
     return autoproxy(token, serializer, manager, authkey,
                      exposed, incref)
 

--- a/sdks/python/apache_beam/utils/multi_process_shared.py
+++ b/sdks/python/apache_beam/utils/multi_process_shared.py
@@ -44,10 +44,17 @@ import fasteners
 # https://stackoverflow.com/questions/46779860/multiprocessing-managers-and-custom-classes
 autoproxy = multiprocessing.managers.AutoProxy
 
-def patched_autoproxy(token, serializer, manager=None, authkey=None,
-          exposed=None, incref=True, manager_owned=True):
-    return autoproxy(token, serializer, manager, authkey,
-                     exposed, incref)
+
+def patched_autoproxy(
+    token,
+    serializer,
+    manager=None,
+    authkey=None,
+    exposed=None,
+    incref=True,
+    manager_owned=True):
+  return autoproxy(token, serializer, manager, authkey, exposed, incref)
+
 
 multiprocessing.managers.AutoProxy = patched_autoproxy
 

--- a/sdks/python/apache_beam/utils/multi_process_shared.py
+++ b/sdks/python/apache_beam/utils/multi_process_shared.py
@@ -24,6 +24,8 @@ on it via rpc.
 
 import logging
 import multiprocessing.managers
+from inspect import signature
+from functools import wraps
 import os
 import tempfile
 import threading
@@ -35,6 +37,19 @@ from typing import Optional
 from typing import TypeVar
 
 import fasteners
+
+# Backup original AutoProxy function
+backup_autoproxy = multiprocessing.managers.AutoProxy
+
+# Defining a new AutoProxy that handles unwanted key argument 'manager_owned'
+def redefined_autoproxy(token, serializer, manager=None, authkey=None,
+          exposed=None, incref=True, manager_owned=True):
+    # Calling original AutoProxy without the unwanted key argument
+    return backup_autoproxy(token, serializer, manager, authkey,
+                     exposed, incref)
+
+# Updating AutoProxy definition in multiprocessing.managers package
+multiprocessing.managers.AutoProxy = redefined_autoproxy
 
 T = TypeVar('T')
 AUTH_KEY = b'mps'
@@ -55,7 +70,7 @@ class _SingletonProxy:
       raise RuntimeError('Entry was released.')
     return self._SingletonProxy_entry.obj.__call__(*args, **kwargs)
 
-  def _SingletonProxy_release(self):
+  def singletonProxy_release(self):
     assert self._SingletonProxy_valid
     self._SingletonProxy_valid = False
 
@@ -68,6 +83,7 @@ class _SingletonProxy:
     # Needed for multiprocessing.managers's proxying.
     dir = self._SingletonProxy_entry.obj.__dir__()
     dir.append('singletonProxy_call__')
+    dir.append('singletonProxy_release')
     return dir
 
 
@@ -92,7 +108,7 @@ class _SingletonEntry:
       return _SingletonProxy(self)
 
   def release(self, proxy):
-    proxy._SingletonProxy_release()
+    proxy.singletonProxy_release()
     with self.lock:
       self.refcount -= 1
       if self.refcount == 0:
@@ -150,6 +166,9 @@ class _AutoProxyWrapper:
 
   def __getattr__(self, name):
     return getattr(self._proxyObject, name)
+
+  def get_auto_proxy_object(self):
+    return self._proxyObject
 
 
 class MultiProcessShared(Generic[T]):
@@ -252,7 +271,7 @@ class MultiProcessShared(Generic[T]):
     return _AutoProxyWrapper(singleton)
 
   def release(self, obj):
-    self._manager.release_singleton(self._tag, obj)
+    self._manager.release_singleton(self._tag, obj.get_auto_proxy_object())
 
   def _create_server(self, address_file):
     # We need to be able to authenticate with both the manager and the process.

--- a/sdks/python/apache_beam/utils/multi_process_shared.py
+++ b/sdks/python/apache_beam/utils/multi_process_shared.py
@@ -40,7 +40,7 @@ import fasteners
 # the kwarg 'manager_owned'. We implement our own backup here to make sure
 # we avoid this problem. More info here:
 # https://stackoverflow.com/questions/46779860/multiprocessing-managers-and-custom-classes
-autoproxy = multiprocessing.managers.AutoProxy
+autoproxy = multiprocessing.managers.AutoProxy  # type: ignore[attr-defined]
 
 
 def patched_autoproxy(
@@ -54,7 +54,7 @@ def patched_autoproxy(
   return autoproxy(token, serializer, manager, authkey, exposed, incref)
 
 
-multiprocessing.managers.AutoProxy = patched_autoproxy
+multiprocessing.managers.AutoProxy = patched_autoproxy  # type: ignore[attr-defined]
 
 T = TypeVar('T')
 AUTH_KEY = b'mps'

--- a/sdks/python/apache_beam/utils/multi_process_shared.py
+++ b/sdks/python/apache_beam/utils/multi_process_shared.py
@@ -38,18 +38,19 @@ from typing import TypeVar
 
 import fasteners
 
-# Backup original AutoProxy function
-backup_autoproxy = multiprocessing.managers.AutoProxy
+# In some python versions, there is a bug where AutoProxy doesn't handle
+# the kwarg 'manager_owned'. We implement our own backup here to make sure
+# we avoid this problem. More info here:
+# https://stackoverflow.com/questions/46779860/multiprocessing-managers-and-custom-classes
+autoproxy = multiprocessing.managers.AutoProxy
 
-# Defining a new AutoProxy that handles unwanted key argument 'manager_owned'
-def redefined_autoproxy(token, serializer, manager=None, authkey=None,
+def patched_autoproxy(token, serializer, manager=None, authkey=None,
           exposed=None, incref=True, manager_owned=True):
     # Calling original AutoProxy without the unwanted key argument
-    return backup_autoproxy(token, serializer, manager, authkey,
+    return autoproxy(token, serializer, manager, authkey,
                      exposed, incref)
 
-# Updating AutoProxy definition in multiprocessing.managers package
-multiprocessing.managers.AutoProxy = redefined_autoproxy
+multiprocessing.managers.AutoProxy = patched_autoproxy
 
 T = TypeVar('T')
 AUTH_KEY = b'mps'

--- a/sdks/python/apache_beam/utils/multi_process_shared_test.py
+++ b/sdks/python/apache_beam/utils/multi_process_shared_test.py
@@ -133,6 +133,38 @@ class MultiProcessSharedTest(unittest.TestCase):
     with self.assertRaisesRegex(Exception, 'released'):
       counter1.get()
 
+  def test_release_always_proxy(self):
+    shared1 = multi_process_shared.MultiProcessShared(
+        Counter, tag='test_release_always_proxy', always_proxy=True)
+    shared2 = multi_process_shared.MultiProcessShared(
+        Counter, tag='test_release_always_proxy', always_proxy=True)
+
+    counter1 = shared1.acquire()
+    counter2 = shared2.acquire()
+    self.assertEqual(counter1.increment(), 1)
+    self.assertEqual(counter2.increment(), 2)
+
+    counter1again = shared1.acquire()
+    self.assertEqual(counter1again.increment(), 3)
+
+    shared1.release(counter1)
+    shared2.release(counter2)
+
+    with self.assertRaisesRegex(Exception, 'released'):
+      counter1.get()
+    with self.assertRaisesRegex(Exception, 'released'):
+      counter2.get()
+
+    self.assertEqual(counter1again.get(), 3)
+
+    shared1.release(counter1again)
+
+    counter1New = shared1.acquire()
+    self.assertEqual(counter1New.get(), 0)
+
+    with self.assertRaisesRegex(Exception, 'released'):
+      counter1.get()
+
 
 if __name__ == '__main__':
   logging.getLogger().setLevel(logging.INFO)

